src/zkevm_specs/evm_circuit/execution/end_block.py

```diff
@@ -85,6 +89,18 @@ def end_block(instruction: Instruction):
             ]
         )
     )
+
+    # a valid withdrawal should have non-zero amount
+    total_withdrawals = FQ(
+        len(
+            [
+                wd_row
+                for wd_row in instruction.tables.withdrawal_table
+                if wd_row.amount.expr() != FQ(0)
+            ]
+        )
+    )
```

Comment:
I looked into how withdrawal is processed in a consensus client and an execution client. And I think we can safely assume the amount is non-zero. If the withdrawal contains zero amount, then the withdrawal shouldn't pass the consensus client check.

- In consensus client. 
  - It checked the withdrawal root from execution payload is the same as the one in the beacon state.
  - It checks the withdrawals all have amounts greater than 0 (Using [greater than sign](https://github.com/ethereum/consensus-specs/blob/dev/specs/capella/beacon-chain.md#is_fully_withdrawable_validator) in both fully and partially withdrawal cases)

```suggestion
    total_withdrawals = FQ(len(instruction.tables.withdrawal_table))
```

Reply:
The purpose of checking non-zero `amount` is to tell valid withdrawals and padding withdrawals. Do you think we need to check another field or adding more comment for this?

Reply:
We should create a tag to indicate if it is a valid withdraw or padding.

Reply:
Wondering what is the criteria of having a tag for padding rows? From [tx_circuit implementation](https://github.com/privacy-scaling-explorations/zkevm-specs/blob/master/src/zkevm_specs/tx_circuit.py#L206), it only checks `address == 0` to determine it's a padding row or not. I just followed tx_circuit implementation.

Reply:
In tx table there are 3 kinds of transactions
1. valid txs
2. invalid txs (check validity of `balance`, `nonce`, and `intrinsic gas`)
3. padding transaction (address == 0)

case 2 is to handle L2 specific scenarios to introduce extra cost when transaction is invalid. In withdrawl we might safely assume all withdrawal transactions passed from validators are valid therefore we just check padding by amount = 0, since withdrawl is "pushed" from consensus client -> EVM, and L2 who used the same codebase should follow similar protocols.
I think we can follow @KimiWu123 current approach first to avoid introduce extra tag to indicate invalid withdraw. 

However I was wondering is there better field to differentiate padding, e.g. withdrawal_index == 0 or validator_index == 0 other than amount?

Reply:
> However I was wondering is there better field to differentiate padding, e.g. withdrawal_index == 0 or validator_index == 0 other than amount?

@hero78119 , I was also thinking to use `withdrawal_index == 0`, but `0` is not a invalid index, the index of first withdawal is `0`. If we don't veriy the history of withdrawals, `0` looks fine with me or we check`withdrawal_index == 0 && amount == 0`? Maybe follow tx circuit to check `address == 0`? any of them is fine with me, I don't have any preference.

Reply:
> The approach now makes sense to me. Can we add a comment like "we use amount=0 for padding withdrawals" for clarity?

fixed in [272535e](https://github.com/privacy-scaling-explorations/zkevm-specs/pull/488/commits/272535eac60f968a161580733890782ff7416c7b)